### PR TITLE
Feat/token all feat (All merged branches)

### PR DESCRIPTION
**Pull request by sumitnitsurat**
- added sidebar component with beneficiary & holder view
- ether.js integration
- dummy oa-token integration with saga

### CODE REVIEW

File: src/components/AssetInfo/AssetInfo.test.tsx [hunk 1]
```diff
@@ -1,11 +1,33 @@
...
+    jest.restoreAllMocks();
+  });
+  it("renders with correct etherscan url", async () => {
+    let wrapper: any;
+    await act(async () => {
+      wrapper = mount(<AssetInfo document={TokenRegistry as SignedDocument} />);
+    });
+    expect(wrapper.find("#asset-info-etherscan-link").text()).toStrictEqual("Manage Asset");
+    expect(wrapper.find("a").prop("href")).toStrictEqual(
+      "https://ropsten.etherscan.io/token/0x48399Fb88bcD031C556F53e93F690EEC07963Af3?a=114183028386037208183595590748001008777078424164408283716893897085552431216953"
```
Comment (yehjxraymond) [inline src/components/AssetInfo/AssetInfo.test.tsx:30]: Yep, so in this instance this component could have been a really small feature branch by itself. Next time can put this in it's on branch.

File: src/components/AssetInfo/AssetInfo.tsx [hunk 2]
```diff
@@ -1,22 +1,77 @@
...
       <a
         href={makeEtherscanTokenURL({ registryAddress, tokenId })}
         id="asset-info-etherscan-link"
         rel="noreferrer noopener"
         target="_blank"
+        onClick={handlerToggleSideBar}
       >
         Manage Asset
       </a>
+      <TokenSideBar
```
Comment (yehjxraymond) [inline src/components/AssetInfo/AssetInfo.tsx]: image

Feel that this will cause the sidebar to be too far deep into the DOM. It's a big component by itself, I will imagine it to be together with the navbar, renderer, etc. Should remove this from AssetInfo, since the "Manage Asset" link doesnt need metamask and all the props that `TokenSidebar` needs. 

Basically we feel that

File: src/components/AssetInfo/TokenSideBar.test.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        approvedBeneficiaryAddress=""
+        handler={() => {}}
+      />
+    );
+    expect(wrapper.find("TokenSideBarRole h4").text()).toStrictEqual("Holder");
+    expect(wrapper.find("TokenSideBarHolder")).toHaveLength(1);
+    expect(wrapper.find("TokenSideBarHolder section")).toHaveLength(1);
+
+    expect(wrapper.find("#sec-transferownership h4").text()).toStrictEqual("Transfer Holdership");
+    expect(wrapper.find("#sec-transferownership button").text()).toStrictEqual("Transfer");
```
Comment (rjchow) [inline src/components/AssetInfo/TokenSideBar.test.tsx:64]: TODO: make issue to review terminology with UXD

File: src/components/AssetInfo/TokenSideBar.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  adminAddress: string;
+  holderAddress: string;
+  beneficiaryAddress: string;
+  approvedBeneficiaryAddress: string;
+  registryAddress?: string;
+  handler: ((event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void) | undefined;
+  isSideBarExpand: boolean;
+}
+
+const TokenSideBar = ({
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBar.tsx:16]: So in our mind we were thinking that the sidebar refer to the actual bar that is shown/hidden. Thinking that a better name for this component will be `TokenSideBarContainer` which has both the sidebar and the expander.

File: src/components/AssetInfo/TokenSideBar.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  adminAddress,
+  holderAddress,
+  beneficiaryAddress,
+  approvedBeneficiaryAddress,
+  registryAddress,
+  handler,
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBar.tsx:22]: Since the container is now self-sufficient to handle it's opened/closed state, we should move the state into this component, instead of passing down from it's parent.

Comment (yehjxraymond) [reply src/components/AssetInfo/TokenSideBar.tsx:22]: also.. `handler` is a bad name, a better function name can be `toggleSidebarVisbility`

File: src/components/AssetInfo/TokenSideBarField.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+interface TokenSideBarFieldProps {
+  id: string;
+  title: string;
+  ctaText: string;
+  ctaStatus?: string;
+  children?: React.ReactNode;
+  handleClick: (e: any) => void;
+}
+
+const TokenSideBarField = ({ id, title, ctaText, ctaStatus, children, handleClick }: TokenSideBarFieldProps) => {
```
Comment (rjchow) [inline src/components/AssetInfo/TokenSideBarField.tsx]: assuming CTA means call-to-action, maybe we should just name ctaText as "label" and status? not too sure why CTA is relevant

File: src/components/AssetInfo/TokenSideBarContent.tsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  adminAddress,
+  beneficiaryAddress,
+  holderAddress,
+  approvedBeneficiaryAddress,
+  registryAddress
+}: TokenSideBarContentProps) => {
+  const [fieldValue, setFieldValue] = useState({ newHolder: "", approvedBeneficiary: "" });
+  const [showActionLoader, toggleActionLoader] = useState(false);
+  const isEqualBeneficiaryAndHolder = adminAddress === holderAddress && adminAddress === beneficiaryAddress;
+  const showHolder = adminAddress === holderAddress || isEqualBeneficiaryAndHolder;
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBarContent.tsx]:
```suggestion
  const showHolder = adminAddress === holderAddress;
```

File: src/components/AssetInfo/TokenSideBarContent.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  const showBeneficiary = adminAddress === beneficiaryAddress && !isEqualBeneficiaryAndHolder;
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBarContent.tsx]:
```suggestion
  const showBeneficiary = (holderAddress !== beneficiaryAddress) && (adminAddress === beneficiaryAddress);
```

File: src/components/AssetInfo/TokenSideBarContent.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+const TokenSideBarContent = ({
+  adminAddress,
+  beneficiaryAddress,
+  holderAddress,
+  approvedBeneficiaryAddress,
+  registryAddress
+}: TokenSideBarContentProps) => {
+  const [fieldValue, setFieldValue] = useState({ newHolder: "", approvedBeneficiary: "" });
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBarContent.tsx:25]: The `TokenSideBarHolder` should handle it's own state of the inputs. Think about how to put ` TokenSideBarHolder` in a storybook, it will not be able to handle input changes. 

I will imagine that we pass in functions such as `handleTransferOwnership(newAddress)` to the `TokenSideBarHolder`.

File: src/components/AssetInfo/TokenSideBarContent.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  const [showActionLoader, toggleActionLoader] = useState(false);
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBarContent.tsx:26]: toggle implies that is doesnt need a param. Will be better to call this setActionLoader

File: src/components/AssetInfo/TokenSideBarContent.tsx [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  };
+
+  const changeBeneficiary = () => {};
+
+  const surrenderDocument = () => {};
+
+  return (
+    <>
+      {(loadUserAddress || showActionLoader) && <div className={css.loader} />}
+      {showNoAccess && <TokenSideBarNoMatch />}
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBarContent.tsx]: Thinking that the visibility should be handled in each of these sections independently. That way, we can leave the big chunk of logic out here.

Comment (yehjxraymond) [reply src/components/AssetInfo/TokenSideBarContent.tsx]: Only when there is a transaction going on, we can hide all the sections...

File: src/components/AssetInfo/TokenSideBarContent.tsx [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+      {showHolder && (
+        <TokenSideBarHolder
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBarContent.tsx:64]: Following the above train of thought, we can break this down into 3 separate component to handle:
1. transferHoldership
2. changeBeneficiary
3. surrenderDocument

File: src/components/AssetInfo/TokenSideBarContent.tsx [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+          newHolder={fieldValue.newHolder}
+          setNewHolder={handleInputChange}
+          transferHoldership={transferHoldership}
+          changeBeneficiary={changeBeneficiary}
+          surrenderDocument={surrenderDocument}
+        />
+      )}
+      {showBeneficiary && (
+        <TokenSideBarBeneficiary
+          setBeneficiary={handleInputChange}
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBarContent.tsx:77]: As above...

File: src/components/AssetInfo/TokenSideBarRole.tsx [hunk 14]
```diff
@@ -0,0 +1,26 @@
+import React from "react";
+
+interface TokenSideBarRoleProps {
+  adminAddress: string;
+  beneficiaryAddress: string;
+  holderAddress: string;
+}
+
+const TokenSideBarRole = ({ adminAddress, beneficiaryAddress, holderAddress }: TokenSideBarRoleProps) => {
+  const getUserRole = () => {
```
Comment (yehjxraymond) [inline src/components/AssetInfo/TokenSideBarRole.tsx]: We can reuse this logic in many place. Let's enum the state and maybe put this in the hook to get the user role directly.

File: src/components/CertificateVerifyBlock/DetailedCertificateVerifyBlock/DetaledCertificateVerifyBlock.test.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
+import React from "react";
```
Comment (yehjxraymond) [inline src/components/CertificateVerifyBlock/DetailedCertificateVerifyBlock/DetaledCertificateVerifyBlock.test.js]: File name is misspelt

File: src/services/token/token.js [hunk 16]
```diff
@@ -5,21 +5,47 @@ import { getLogger } from "../../utils/logger";
```
Comment (rjchow) [inline src/services/token/token.js:5]: TODO: let's refactor this to be encapsulated so that we don't introduce bugs down the line

File: src/components/CertificateVerifyBlock/DetailedCertificateVerifyBlock/DetaledCertificateVerifyBlock.test.js [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+import { shallow } from "enzyme";
+import { act } from "react-dom/test-utils";
+import DetailedCertificateVerifyBlock from ".";
+import ROPSTEN from "../../HomePageContent/Ropsten-Demo.json";
+import TokenRegistry from "../../../test/fixture/tokenRegistry.json";
+
+describe("detailedCertificateVerifyBlock", () => {
+  it("render document is token", async () => {
```
Comment (yehjxraymond) [inline src/components/CertificateVerifyBlock/DetailedCertificateVerifyBlock/DetaledCertificateVerifyBlock.test.js]:
```suggestion
  it("renders if document is a token", async () => {
```

File: src/components/CertificateVerifyBlock/DetailedCertificateVerifyBlock/DetaledCertificateVerifyBlock.test.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
+import React from "react";
+import { shallow } from "enzyme";
+import { act } from "react-dom/test-utils";
+import DetailedCertificateVerifyBlock from ".";
+import ROPSTEN from "../../HomePageContent/Ropsten-Demo.json";
+import TokenRegistry from "../../../test/fixture/tokenRegistry.json";
```
Comment (yehjxraymond) [inline src/components/CertificateVerifyBlock/DetailedCertificateVerifyBlock/DetaledCertificateVerifyBlock.test.js]:
```suggestion
import sampleToken from "../../../test/fixture/sample-token.json";
```

Have created an issue to rename the file, we can do it here or in another PR

File: src/components/CertificateViewer.js [hunk 19]
```diff
@@ -25,19 +24,13 @@ const renderVerifyBlock = props => (
...
-};
-
 const renderHeaderBlock = props => {
   const renderedVerifyBlock = renderVerifyBlock(props);
   return (
     <div className={`container-fluid ${styles["pd-0"]} ${styles.container}`}>
       <div className="row">
         <div>{renderedVerifyBlock}</div>
-        <div>{renderAssetInfo(props)}</div>
+        <AssetInfo document={props.document} />
```
Comment (yehjxraymond) [inline src/components/CertificateViewer.js:33]: Comment above to move this out of here. Can rename AssetInfo to something like `ManageAssetToggle`

File: src/services/etherjs/accounts.ts [hunk 20]
```diff
@@ -0,0 +1,12 @@
+import { ethers } from "ethers";
+
+// Wrap provider.listAccounts in a Promise
+export const getAccounts = (provider: ethers.providers.JsonRpcProvider) =>
```
Comment (yehjxraymond) [inline src/services/etherjs/accounts.ts:4]: use await async

File: src/reducers/admin.js [hunk 21]
```diff
@@ -0,0 +1,54 @@
+export const initialState = {
+  adminAddress: "",
```
Comment (yehjxraymond) [inline src/reducers/admin.js:2]: userWalletAddress

File: src/sagas/admin.js [hunk 22]
```diff
@@ -0,0 +1,30 @@
+import { put, takeEvery } from "redux-saga/effects";
+import { types } from "../reducers/admin";
+
+import getAccounts from "../services/etherjs/accounts";
```
Comment (yehjxraymond) [inline src/sagas/admin.js:4]: can we export this from ethersjs too?

File: src/sagas/token.js [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+
+function* checkIfTitleEscrow(document) {
+  try {
+    const isTitleEscrow = yield call(isEscrowContract, document);
+    if (!isTitleEscrow) throw new Error("Document owner is not a escrow contract");
+    yield put(setIsEscrowContractSuccess());
+    return isTitleEscrow;
+  } catch (e) {
+    error(`checkIfTitleEscrow: ${JSON.stringify(e)}`);
+    yield put(setIsEscrowContractError());
```
Comment (yehjxraymond) [inline src/sagas/token.js:36]: put error as a parameter to setIsEscrowContractError

File: src/sagas/token.js [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+  } catch (e) {
+    yield put(getTokenUserAddressError(e.message));
+  }
+}
+
+export function* initializeToken() {
+  try {
+    const document = yield select(getCertificate);
+    const { provider, signer } = yield getProvider();
+    //trace(`Web3 provider: ${JSON.stringify(provider)}`);
```
Comment (yehjxraymond) [inline src/sagas/token.js]: remove this

File: src/sagas/token.js [hunk 25]
```diff
@@ -0,0 +1,93 @@
...
+    yield put(initializeTokenFailure(e.message));
+  }
+}
+
+export function* transferOwnership({ payload }) {
+  try {
+    const document = yield select(getCertificate);
+    const { newTokenOwner } = payload;
+
+    const transferStatus = yield transferTokenOwnership(document, newTokenOwner);
```
Comment (yehjxraymond) [inline src/sagas/token.js:80]: the function actually returns a string, transaction hash after the promise resolves

File: src/components/AssetInfo/TokenSideBar.test.tsx (hunk 3 above)
Comment (sumitnitsurat) [reply src/components/AssetInfo/TokenSideBar.test.tsx:64]: Boon made an issue for this.

File: src/components/AssetInfo/TokenSideBarField.tsx (hunk 6 above)
Comment (simboonlong) [reply src/components/AssetInfo/TokenSideBarField.tsx]: ok, updated the text.

File: src/components/AssetInfo/TokenSideBarRole.tsx (hunk 14 above)
Comment (simboonlong) [reply src/components/AssetInfo/TokenSideBarRole.tsx]: Updated.

File: src/components/CertificateVerifyBlock/DetailedCertificateVerifyBlock/DetaledCertificateVerifyBlock.test.js (hunk 18 above)
Comment (simboonlong) [reply src/components/CertificateVerifyBlock/DetailedCertificateVerifyBlock/DetaledCertificateVerifyBlock.test.js]: Renamed the file and respective imports.

File: src/components/AssetInfo/TokenSideBarRole.tsx (hunk 14 above)
Comment (sumitnitsurat) [reply src/components/AssetInfo/TokenSideBarRole.tsx]: refacotred it.

**Review by rjchow — APPROVED** [review]
